### PR TITLE
[new recipe, algo, ci, doc]{feat} DisCO algorithm implementation with related recipe, CI tests and docs updates

### PR DESCRIPTION
**Pull request by GangLii**
### What does this PR do?

This PR introduces a novel RL algorithm named Discriminative Constrained Optimization(DisCO) for LLM Reasoning with the Verl framework. It trains models by **increasing scores for positive answers while decreasing those for negatives**, enabling: (a) *No Early Entropy Collapse*, (b) *Faster convergence*, (c) *More stable training*, (d) *Handles sparse rewards*.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

CI tests have been added to test the new algorithm. Moreover, additional experiments have been conducted to verify the implementation.
image

### API and Usage Example
```bash
cd verl # Repo root
bash recipe/disco/run_disco_1.5b.sh # or other scripts
```

### Design & Code Changes

Code changes include:

* Algorithm implementation: 
     * `verl/worker/config.actor.py`: add config support for DisCO algorithm
     *  `verl/trainer/ppo/core_algos.py` add advantage function and policy loss for DisCO algorithm
* Recipe:
     * `recips/disco`: including necessary code and configs for running the experiments shown in the [paper](https://arxiv.org/abs/2505.12366).
* CI tests:
    * `tests/specials_e2e/run_disco.sh`
    * `.github/workflows/e2e_disco.yml`
* docs:
    * `docs/algo/disco.md`
    * `docs/index.rst`

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -1235,6 +1278,131 @@ def compute_policy_loss_geo_mean(
...
+
+    ### group scores based on global_uid, we assume each question has same number of responses
+    sorted_uid, indices = global_uid.sort()
+    sorted_scores = global_scores[indices]
+    sorted_rewards = global_rewards[indices]
+
+    num_questions = global_uid.unique().numel()
+    num_responses_per_question = global_uid.size(0) // num_questions
+    grouped_scores = sorted_scores.view(num_questions, num_responses_per_question)
+    grouped_rewards = sorted_rewards.view(num_questions, num_responses_per_question)
```
[the remark this answers is not in the document]
Comment (GangLii) [reply verl/trainer/ppo/core_algos.py:1363]: In the current version, the code is intentionally designed like this to make sure each question has the same number of responses in one forward.

File: recipe/disco/reward/deepscaler_reward.py [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+    if answer is None:
+        return None
+    answer = answer.strip()
+    try:
+        # Remove enclosing `\text{}`.
+        m = re.search("^\\\\text\{(?P<text>.+?)\}$", answer)
+        if m is not None:
+            answer = m.group("text").strip()
+        return _strip_string(answer)
+    except:  # noqa: E722
```
[the remark this answers is not in the document]
Comment (GangLii) [reply recipe/disco/reward/deepscaler_reward.py:129]: As the code is borrowed from other researchers, we prefer to keep it untouched to ensure it functions correctly.

File: verl/trainer/ppo/core_algos.py [hunk 3]
```diff
@@ -1235,6 +1278,131 @@ def compute_policy_loss_geo_mean(
...
+
+        # Subtract max, exponentiate, and apply mask
+        neg_exp = torch.exp(((grouped_scores / tau) - neg_max.detach()) * neg_mask) * neg_mask
+        neg_sum_exp = neg_exp.sum(dim=-1, keepdim=True)
+
+        neg_logmeanexp = neg_sum_exp / (neg_sum_exp.detach() + torch.finfo(neg_sum_exp.dtype).eps)
+
+        pg_losses = ((grouped_scores - tau * neg_logmeanexp) * pos_mask).sum(dim=1, keepdim=True) / pos_mask.sum(
+            dim=1, keepdim=True
+        )
```
[the remark this answers is not in the document]
Comment (GangLii) [reply verl/trainer/ppo/core_algos.py:1391]: The gradient of the current implementation works as expected and was verified, and the loss values are not important in the experiments.

File: docs/algo/disco.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+<p align="center"><img alt="Comparison with baselines on 1.5B model" src="https://github.com/Optimization-AI/DisCO/blob/main/assets/1p5model.png" width="800"/></p>
+
+
+Comparison with baseline models and baseline methods for fine-tuning 7B models. Methods in the bottom area are all for fine-tuning  DeepSeek-R1-Distill-Qwen-7B model on the the same DeepScalaR dataset.
+
+<p align="center"><img alt="Comparison with baselines on 7B model" src="https://github.com/Optimization-AI/DisCO/blob/main/assets/7Bmodel.png" width="800"/></p>
+
+Training dynamics of different methods: left two are for fine-tuning 1.5B model and right two are for fine-tuning 7B model. (a), (c) plot the training reward (averaged over generated outputs for questions used in each step) vs the number of training steps; (b), (d) plot the generation entropy vs training steps.
+
+<p align="center"><img alt="Training Dynamics" src="https://github.com/Optimization-AI/DisCO/blob/main/assets/training-dyanmics.png" width="800"/></p>
```
[the remark this answers is not in the document]
Comment (GangLii) [reply docs/algo/disco.md]: The typo is fixed as suggested.

File: recipe/disco/README.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+<p align="center"><img alt="Comparison with baselines on 1.5B model" src="https://github.com/Optimization-AI/DisCO/blob/main/assets/1p5model.png" width="800"/></p>
+
+
+Comparison with baseline models and baseline methods for fine-tuning 7B models. Methods in the bottom area are all for fine-tuning  DeepSeek-R1-Distill-Qwen-7B model on the the same DeepScalaR dataset.
+
+<p align="center"><img alt="Comparison with baselines on 7B model" src="https://github.com/Optimization-AI/DisCO/blob/main/assets/7Bmodel.png" width="800"/></p>
+
+Training dynamics of different methods: left two are for fine-tuning 1.5B model and right two are for fine-tuning 7B model. (a), (c) plot the training reward (averaged over generated outputs for questions used in each step) vs the number of training steps; (b), (d) plot the generation entropy vs training steps.
+
+<p align="center"><img alt="Training Dynamics" src="https://github.com/Optimization-AI/DisCO/blob/main/assets/training-dyanmics.png" width="800"/></p>
```
[the remark this answers is not in the document]
Comment (GangLii) [reply recipe/disco/README.md]: The typo is fixed as suggested.

File: verl/workers/config/actor.py [hunk 6]
```diff
@@ -35,12 +35,16 @@ class PolicyLossConfig(BaseConfig):
...
 
     Args:
-        loss_mode (str): Loss function mode. Options: 'vanilla', 'clip-cov', 'kl-cov', 'gpg'.
+        loss_mode (str): Loss function mode. Options: 'vanilla', 'clip-cov', 'kl-cov', 'gpg', 'disco'.
         clip_cov_ratio (float): Ratio of tokens to be clipped for clip-cov loss.
         clip_cov_lb (float): Lower bound for clip-cov loss.
         clip_cov_ub (float): Upper bound for clip-cov loss.
         kl_cov_ratio (float): Ratio of tokens to be applied KL penalty for kl-cov loss.
         ppo_kl_coef (float): KL divergence penalty coefficient.
+        score_func (str):  Score function for disco loss. Options: 'logL', 'Lratio'.
```
Comment (vermouth1992) [inline verl/workers/config/actor.py:44]: I guess we will need to refactor this to put algorithm specific parameter into a separate config. Let me do this on Monday

Comment (GangLii) [reply verl/workers/config/actor.py:44]: Hi @vermouth1992 , I'd like to follow up on refactoring the parameter config. Please feel free to let me know if there is anything I can help with.

Comment (GangLii) [reply verl/workers/config/actor.py:44]: Hi @vermouth1992, I’ve updated the code to address the CI test errors reported in the workflows. Could you please re-trigger the testing workflow? Let me know if there’s anything else I can help with. Thank you!

**Comment by GangLii** [comment]
Hi @eric-haibin-lin @vermouth1992 @tongyx361 @PeterSH6 @techkang, we are excited to share that our DisCO work was accepted to NeurIPS 2025. 

As we would like to help researchers in the community utilize our work more easily via the powerful and convenient Verl framework, could you help review the code and let me know if there are any changes needed? Thank you!
